includes/class-wc-payments-order-service.php

```diff
@@ -0,0 +1,706 @@
+		$this->add_capture_expired_note( $order, $intent_id, $charge_id );
+		$this->complete_order_processing( $order, $intent_status );
+	}
+
+	/**
+	 * Updates an order to cancelled status, while adding a note with a link to the transaction.
+	 *
+	 * @param WC_Order $order         Order object.
+	 * @param string   $intent_id     The ID of the intent associated with this order.
+	 * @param string   $intent_status The status of the intent related to this order.
+	 * @param string   $charge_id     The charge ID related to the intent/order.
+	 *
+	 * @return void
+	 */
+	public function mark_payment_capture_cancelled( $order, $intent_id, $intent_status, $charge_id ) {
+		if ( ! $this->order_prepared_for_processing( $order, $intent_id ) ) {
+			return;
+		}
+
+		$this->update_order_status( $order, 'cancelled' );
+		$this->add_capture_cancelled_note( $order, $intent_id, $charge_id );
+		$this->complete_order_processing( $order, $intent_status );
+	}
+
+	/**
+	 * Updates the order to on-hold status and adds a note about the dispute.
+	 *
+	 * @param WC_Order $order      Order object.
+	 * @param string   $dispute_id The ID of the dispute associated with this order.
+	 * @param string   $reason     The reason for the dispute.
+	 *
+	 * @return void
+	 */
+	public function mark_payment_dispute_created( $order, $dispute_id, $reason ) {
+		if ( ! is_a( $order, 'WC_Order' ) ) {
+			return;
+		}
+
+		$this->update_order_status( $order, 'on-hold' );
+		$this->add_dispute_created_note( $order, $dispute_id, $reason );
+		$order->save();
+	}
+
+	/**
+	 * Updates the order status based on dispute status and adds a note about the dispute.
+	 *
+	 * @param WC_Order $order      Order object.
+	 * @param string   $dispute_id The ID of the dispute associated with this order.
+	 * @param string   $status     The status of the dispute.
+	 *
+	 * @return void
+	 */
+	public function mark_payment_dispute_closed( $order, $dispute_id, $status ) {
+		if ( ! is_a( $order, 'WC_Order' ) ) {
+			return;
+		}
+
+		if ( 'lost' === $status ) {
+			wc_create_refund(
+				[
+					'amount'     => $order->get_total(),
+					'reason'     => __( 'Dispute lost.', 'woocommerce-payments' ),
+					'order_id'   => $order->get_id(),
+					'line_items' => $order->get_items(),
+				]
+			);
+		} else {
+			$this->update_order_status( $order, 'completed' );
+		}
+
+		$this->add_dispute_closed_note( $order, $dispute_id, $status );
+		$order->save();
+	}
+
+	/**
+	 * Updates a terminal order to completed status, while adding a note with a link to the transaction.
+	 *
+	 * @param WC_Order $order         Order object.
+	 * @param string   $intent_id     The ID of the intent associated with this order.
+	 * @param string   $intent_status The status of the intent related to this order.
+	 * @param string   $charge_id     The charge ID related to the intent/order.
+	 *
+	 * @return void
+	 */
+	public function mark_terminal_payment_completed( $order, $intent_id, $intent_status, $charge_id ) {
```

Comment:
I wonder whether it's possible to use `mark_payment_completed` here, rather than having two functions. Perhaps by passing through a bool flag to `mark_payment_completed` such as `$terminal = false` and then setting it to true if it is a terminal payment.

Is there a significant difference between how we mark a terminal payment as completed vs a normal payment? If there is, I think it's fine to have the two functions of course.

Reply:
Basically all `mark_terminal_payment_completed` is doing at this point is moving the order to the `completed` status. It also updates the intent on the order, like the other methods in the service does, but I believe that's already done in the code for the terminal transaction and in `mark_payment_completed`.

When it's called, [it has already stepped through the process](https://github.com/Automattic/woocommerce-payments/blob/add/order-service/includes/admin/class-wc-rest-payments-orders-controller.php#L146-L151) of calling [`update_order_status_from_intent`](https://github.com/Automattic/woocommerce-payments/blob/add/order-service/includes/class-wc-payment-gateway-wcpay.php#L1288), which has called `mark_payment_completed`. `mark_payment_completed` adds the note and moves the order to processing if it's a physical product, and since the order is supposedly in person, it should be moved to completed assuming the merchant is physically selling the item to the customer.

Since all it's doing is moving the order to the `completed` status, it's not needed, I was just trying to consolidate the calls to `update_status` for orders into this service.

If there was a better/standard intent object being used between all of these transactions, we could have a flag within that to say "this is a terminal transaction" and it could be marked as completed rather than processing initially, but that's something for a future refactor. I went down so many rabbit holes with trying to make things work better, but things like multiple intent types and using arrays and objects as intents becomes a huge issue.

Reply:
Thanks for the detailed explanation! I agree that we should look at this when we have better models to pass around our code, it works well for now and this PR is laying the groundwork for future refactors anyway.